[docs][guide] writing a multi-asset backed integration (#22910)
dagster-io/dagster: docs/content/guides/integrations/writing-a-multi-asset-decorator-integration.mdx

```

Here we define a function that returns a `multi_asset` function. The `multi_asset` function is a decorator itself, so this allows us to customize the behavior of `multi_asset` and create a new decorator of our own.

We load our replication project, and then iterate over the tables that were defined in the YAML file. Those are then used to create a list of `AssetSpec` objects that are passed to the `specs` parameter. This is what allows us to define the assets that are visible in the Dagster UI. However, we have not yet shown how to perform the actual execution of our replication function — that's what we will go over next!

```python
def custom_replication_assets(
    *,
    replication_project: ReplicationProject,
    name: Optional[str] = None,
    group_name: Optional[str] = None,
) -> Callable[[Callable[..., Any]], AssetsDefinition]:
    project = replication_project.load()

    return multi_asset(
        name=name,
        group_name=group_name,
        specs=[
            AssetSpec(
                key=table.get("name"),
            )
            for table in project.get("tables")
        ],
    )

> ```

Reviewer: Admittedly I'm not the target audience for this guide, but where should the code described here actually go? It's not clear to me how this works with the code snippets that come after it, or what my code should look like after I've completed the guide.

> ```

Author: Your ordering is much more intuitive. Thank you.

> Thinking back to our previous example on how decorators work, they allow us to wrap a function that performs some operation. In the case of our `multi_asset` we are able to define `AssetSpec` objects for our project, and the actual processing that takes place will be in the body of the function we decorate.

Reviewer: ```suggestion
Recall that decorators allow us to wrap a function that performs some operation. In the case of our `multi_asset`, we defined <PyObject object="AssetSpec" /> objects for our project and the actual processing that takes place will be in the body of the decorated function.
```

Also, did we define `AssetSpec` objects for the project? Or for the tables in the project?

Author: Good call, it was the tables.

In this function, we will perform the replication, and then yield `AssetMaterialization` objects indicating that the replication was successful for a given table.

```python
from dagster import AssetExecutionContext

replication_project_path = "replication.yaml"
replication_project = ReplicationProject(replication_project_path)

@custom_replication_assets(
    replication_project=replication_project,
    name="my_custom_replication_assets",
    group_name="replication",
)
def my_assets(context: AssetExecutionContext):
    results = replicate(Path(replication_project_path))
    for table in results:
        if table.get("status") == "SUCCESS":
            yield AssetMaterialization(asset_key=str(table.get("name")), metadata=table)
```

> You may have noticed, however, that there are a couple of limitations to this approach. Most importantly, we have not encapsulated the logic for replicating tables. Meaning, end-users who make use of the `custom_replication_assets` decorator would be responsible for yielding asset materializations themselves. Second, there is no way for the user to customize the attributes of the asset.

Reviewer: ```suggestion
There are a few limitations to this approach:

- **We have not encapsulated the logic for replicating tables.** This means that users who use the `custom_replication_assets` decorator would be responsible for yielding asset materializations themselves. 
- **Users can't customize the attributes of the asset**.
```

Suggesting we cut out some unneeded words and format this as a list to make it easier to read.

> For the first limitation, we can resolve them by refactoring the code in the body of our asset function into a `Resource`.

Reviewer: ```suggestion
For the first limitation, we can resolve this by refactoring the code in the body of our asset function into a Dagster resource.
```

- Nit - the limitation is singular, so the pronoun used should be singular
- We're just talking about a regular resource, right? Seeing `Resource` makes me think that's exactly what I'll see in code.

---

> ## Step 3: Moving our replication logic into a Resource

Reviewer: ```suggestion
## Step 3: Moving the replication logic into a resource
```

The `our` here strikes me as a little odd, and resource doesn't need to be capitalized

> To accomplish this, we will extend the `ConfigurableResource` object to create our own custom resource. Then, we will define a `run` method that will perform the replication operation.

Reviewer: ```suggestion
To accomplish this, we will extend the <PyObject object="ConfigurableResource" /> object to create a custom resource. Then, we will define a `run` method that will perform the replication operation:
```

```python
from dagster import ConfigurableResource
from dagster._annotations import public

class ReplicationResource(ConfigurableResource):
    @public
    def run(
        self, replication_project: ReplicationProject
    ) -> Iterator[AssetMaterialization]:
        results = replicate(Path(replication_project.replication_configuration_yaml))
        for table in results:
            if table.get("status") == "SUCCESS":
                # NOTE: this assumes that the table name is the same as the asset key
                yield AssetMaterialization(
                    asset_key=str(table.get("name")), metadata=table
                )
```

> Now, we can refactor our `custom_replication_assets` instance to use this resource.

Reviewer: ```suggestion
Now, we can refactor our `custom_replication_assets` instance to use this resource:
```

```python
@custom_replication_assets(
    replication_project=replication_project,
    name="my_custom_replication_assets",
    group_name="replication",
)
def my_assets(replication_resource: ReplicationProject):
    replication_resource.run(replication_project)
```

---

## Step 4: Using translators

> Previously we mentioned that there was no way for an end-user to customize the asset definitions, for example, changing the key of the asset. The recommended way to accomplish this is through the use of a translator class.

Reviewer: ```suggestion
At the end of [Step 2](ADD-LINK-HERE), we mentioned end users couldn't customize asset attributes such as changing the key of the asset. Using a translator class is the recommended way to accomplish this.
```

Suggesting we cite and link to where we said this. Also, should it be asset `definitions` or `attributes`? The copy in Step 2 that says this uses `attributes` - I think we should be consistent to prevent confusion.

> This class provides end-users with a way to override the translation between a configuration file, and how that maps to concepts in Dagster.

Reviewer: ```suggestion
This class provides end users with a way to override the translation between a configuration file, and how that maps to concepts in Dagster.
```

- `end-users` doesn't need to be hyphenated
- I'm not sure what `how that maps to concepts in Dagster` means?

Author: I've expanded on this explanation.
> At the end of [Step 2](#step-2-implementation), we mentioned that end users were unable to customize asset attributes, like the asset key, generated by our decorator. Translator classes are the recommended way of defining this logic, and they provide users with the option to override the default methods used to convert a concept from your tool (e.g. a table name) to the corresponding concept in Dagster (e.g. asset key).

> To start, we will define a translator method to map the table specification to a Dagster asset key associated, however, this approach can be mapped to many other attributes such as: dependencies, group name, metadata, and more. You can reference the implementation for other integrations for a full example. It is important to expose all use-cases when defining the parameters passed to the asset decorator, along with methods implemented on the translator.

Reviewer: ```suggestion
To start, we will define a translator method to map the table specification to a Dagster asset key. This approach, however, can be mapped to other attributes like dependencies, group names, metadata, and more. **Note**: When defining the parameters passed to the asset decorator, it's important to expose all use cases along with the methods implemented on the translator.
```

- Suggesting we break up that first sentence to make it easier to read
- About this guy:

  > You can reference the implementation for other integrations for a full example.

   Having this in the middle of this paragraph feels odd, and a little bit like we're telling the user to go find it themselves. Should they refer to the links included in the intro?
- Also suggesting we format the last sentence as `**Note**:` to make it stand out more. I'm not sure if my wording here is totally correct + what you intended, though

```python
from dagster import AssetKey, _check as check

from dataclasses import dataclass

@dataclass
class ReplicationTranslator:
    @public
    def get_asset_key(self, table_definition: Mapping[str, str]) -> AssetKey:
        return AssetKey(str(table_definition.get("name")))
```

> Next, we can update our `custom_replication_assets` to use the translator when defining the `key` on the `AssetSpec`. You will note, that we have taken this opportunity to also include the replication project, and the translator instance on the `AssetSpec` metadata. This is a workaround that we tend to employ in this approach, making it possible to define these objects once, and then access them on the context of our asset.

Reviewer: ```suggestion
Next, we'll update `custom_replication_assets` to use the translator when defining the `key` on the <PyObject object="AssetSpec" />. Note that we took this opportunity to also include the replication project and translator instance on the `AssetSpec` metadata. This is a workaround that we tend to employ in this approach, as it makes it possible to define these objects once and then access them on the context of our asset.
```

- Suggesting we replace `can` in the first sentence - to me it sounded like we could do this, and we may or may not do it next
- Adding `PyObject`
- Suggesting a few wording changes to tighten things up

```python
def custom_replication_assets(
    *,
    replication_project: ReplicationProject,
    name: Optional[str] = None,
    group_name: Optional[str] = None,
    translator: Optional[ReplicationProject] = None,
) -> Callable[[Callable[..., Any]], AssetsDefinition]:
    project = replication_project.load()

    translator = (
        check.opt_inst_param(translator, "translator", ReplicationTranslator)
        or ReplicationTranslator()
    )

    return multi_asset(
        name=name,
        group_name=group_name,
        specs=[
            AssetSpec(
                key=translator.get_asset_key(table),
                metadata={
                    "replication_project": project,
                    "replication_translator": translator,
                },
            )
            for table in project.get("tables")
        ],
    )
```

> Finally, we have to update our resource to use the translator, and project that was provided on the metadata. We are using the `check` method provided by `dagster._check` to ensure that the type of the object is appropriate as we retrieve it from the metadata.

Reviewer: ```suggestion
Finally, we have to update our resource to use the translator and project provided in the metadata. We are using the `check` method provided by `dagster._check` to ensure that the type of the object is appropriate as we retrieve it from the metadata.
```

> Now, we can use the same `translator.get_asset_key` when yielding the asset materialization, thus ensuring that our asset declarations match our asset materializations.

Reviewer: ```suggestion
Now, we can use the same `translator.get_asset_key` when yielding the asset materialization, thus ensuring that our asset declarations match our asset materializations:
```

```python
class ReplicationResource(ConfigurableResource):
    @public
    def run(self, context: AssetExecutionContext) -> Iterator[AssetMaterialization]:
        metadata_by_key = context.assets_def.metadata_by_key
        first_asset_metadata = next(iter(metadata_by_key.values()))

        project = check.inst(
            first_asset_metadata.get("replication_project"),
            ReplicationProject,
        )

        translator = check.inst(
            first_asset_metadata.get("replication_translator"),
            ReplicationTranslator,
        )

        results = replicate(Path(project.replication_configuration_yaml))
        for table in results:
            if table.get("status") == "SUCCESS":
                yield AssetMaterialization(
                    asset_key=translator.get_asset_key(table), metadata=table
                )
```

---

> ## Conclusion

Reviewer: What should the code look like in its entirety? I'm having trouble getting the full picture here.

Author: I've added the full code at the bottom of the conclusion - let me know if this works.

In this guide we walked through how to define a custom multi-asset decorator, a resource for encapsulating tool logic, and a translator for defining the logic to translate a specification to Dagster concepts.

Defining integrations with this approach aligns nicely with the overall development paradigm of Dagster, and is suitable for tools that generate many assets.

> ## References

Reviewer: ```suggestion
---

## Related
```

<ArticleList>
  <ArticleListItem
    title="Unlocking Flexible Pipelines: Customizing the Asset Decorator"
    href="https://dagster.io/blog/unlocking-flexible-pipelines-customizing-asset-decorator"
  ></ArticleListItem>
  <ArticleListItem
    title="dbt multi-asset integration example"

>     href="/integrations/dbt"

Reviewer: This is just pointing to the integration page for dbt, not to the actual code. Idk, this strikes me as a little odd + misleading.

Author: Added links to source code, good call

When working in the Dagster ecosystem, you may have noticed that decorators are frequently used. For example, assets, jobs, and ops use decorators. If you have a service that produces many assets, it's possible to define it as a multi-asset decorator — offering a consistent and intuitive developer experience to existing Dagster APIs.

In the context of Dagster, decorators are helpful because they often wrap some form of processing. For example, when writing an asset, you define your processing code and then annotate the function with the <PyObject object="asset" decorator /> decorator. Then, the internal Dagster code can register the asset, assign metadata, pass in context data, or perform any other variety of operations that are required to integrate your asset code with the Dagster platform.

In this guide, you'll learn how to develop a multi-asset integration for a hypothetical replication tool.

---

## Prerequisites

To follow the steps in this guide, you'll need:

- Familiarity with Dagster
- An understanding of Python decorators — [Real Python's Primer on Python Decorators](https://realpython.com/primer-on-python-decorators/) is a fantastic introduction

---

> 

Reviewer 2: feel free to push back, and we can address it in the jobs-to-be-done world

i wonder if we could add a "when to model your integration in this way" section, where we can explain the whys (e.g. composability), map pattern to category like "ETL", "BI" and also list example implementations like you did on the "Approaches to writing integrations"

Author: I do like that idea - however, let's see if we can merge this first, and then make improvements.

Much of this may be rewritten or scrapped as a part of the docs overhaul. If we can merge this, then we can consider restructuring in the docathon if that works for you.

```python
from dagster import AssetExecutionContext

replication_project_path = "replication.yaml"
replication_project = ReplicationProject(replication_project_path)

@custom_replication_assets(
    replication_project=replication_project,
    name="my_custom_replication_assets",
    group_name="replication",
)
def my_assets(context: AssetExecutionContext):
    results = replicate(Path(replication_project_path))
    for table in results:
        if table.get("status") == "SUCCESS":
            yield AssetMaterialization(asset_key=str(table.get("name")), metadata=table)
```

There are a few limitations to this approach:

- **We have not encapsulated the logic for replicating tables.** This means that users who use the `custom_replication_assets` decorator would be responsible for yielding asset materializations themselves.
- **Users can't customize the attributes of the asset**.

For the first limitation, we can resolve this by refactoring the code in the body of our asset function into a Dagster resource.

---

## Step 3: Moving the replication logic into a resource

> 

Reviewer 2: should we mention the whys? as in, pluggability configurability and reusability

Author: I've included a sentence on this, thanks!
